internal/proxy/deltaproxycache.go

```diff
@@ -77,80 +105,149 @@ func DeltaProxyCacheRequest(r *Request, w http.ResponseWriter, client Client, ca
-					cacheStatus = "rmiss"
-				}
-
 				nts.SetExtents([]timeseries.Extent{*e})
-				metrics.ProxyRequestStatus.WithLabelValues(req.OriginName, req.OriginType, req.HTTPMethod, cacheStatus, strconv.Itoa(resp.StatusCode), req.URL.Path).Inc()
-				metrics.ProxyRequestDuration.WithLabelValues(req.OriginName, req.OriginType, req.HTTPMethod, cacheStatus, strconv.Itoa(resp.StatusCode), req.URL.Path).Observe(elapsed.Seconds())
 				appendLock.Lock()
 				defer appendLock.Unlock()
 
 				mts = append(mts, nts)
 				if rh == nil {
+					// Use the response headers from the first successful delta request to complete as our downstream response headers
 					rh = resp.Header
 				}
 			}
 		}(&missRanges[i], req)
 	}
 
-	// TODO: Fast Forward Here, Another wg.Add(1), go func, defer wg.Done(), fetch, Matrix from Vector, lock, defer unlock, append to mts
+	var ffts timeseries.Timeseries
+	var ffex timeseries.Extent
+	// Only fast forward if configured and the user request is for the absolute latest datapoint
+	if (!cfg.FastForwardDisable) && (trq.Extent.End == normalizedNow.Extent.End) {
+		ffURL, err := client.FastForwardURL(r)
+		if ffURL.Scheme != "" && err == nil {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				req := r.Copy()
+				req.URL = ffURL
+				body, resp := FetchViaObjectProxyCache(req, client, cache, 15, false, true)
+				if resp.StatusCode == http.StatusOK && len(body) > 0 {
+
+					ffts, err = client.UnmarshalInstantaneous(body)
+					if err != nil {
+						log.Error("proxy object unmarshaling failed", log.Pairs{"body": string(body)})
+						return
+					}
+
+					x := ffts.Extents()
+					if len(x) > 0 {
+						ffex = x[0]
+					}
+
+				}
+			}()
+		}
+	}
+
 	wg.Wait()
 
 	// Merge the new delta timeseries into the cached timeseries
-	cts.Merge(mts...)
+	if len(mts) > 0 {
+		cts.Merge(true, mts...)
+	}
 
 	// Get the Request Object, Cropped down from the full Cache
-	rdata, err := client.MarshalTimeseries(cts.Crop(trq.Extent))
+	rts := cts.Crop(trq.Extent)
 
-	wg.Add(1)
-	// Write the newly-merged object back to the cache
-	go func() {
-		// Crop the Cached Object down to the Sample Age Retention Policy before storing
-		re := timeseries.Extent{End: time.Now(), Start: time.Now().Add(-time.Duration(client.Configuration().MaxValueAgeSecs) * time.Second)}
-		cts.Crop(re)
-		cdata, err := client.MarshalTimeseries(cts)
-		if err != nil {
-			return
-		}
-		cacheData.Body = cdata
-		WriteCache(cache, key, cacheData, ttl)
-		wg.Done()
-	}()
+	// Merge Fast Forward data if present. This must be done after the Downstream Crop since
+	// the cropped extent was normalized to stepboundaries and would remove fast forward data
+	if !cfg.FastForwardDisable && ffts != nil && ffex.End.After(trq.Extent.End) {
+		rts.Merge(false, ffts)
+	}
+	rdata, err := client.MarshalTimeseries(rts)
+
+	// Don't write the cache unless it has chnaged
+	if cacheStatus != crHit {
+		wg.Add(1)
+		// Write the newly-merged object back to the cache
+		go func() {
```

Comment:
Is there any chance this will block forever? Does having a timeout on this make sense?